Rest API: add /revisions endpoint for global styles

ramonjd opened the pull request:
The following Global Revisions UI was added to Gutenberg in:

- https://github.com/WordPress/gutenberg/pull/50089

To support this, an endpoint (/wp/v2/global-styles/revisions) that returns revisions for the global styles custom post was added and updated in:

- https://github.com/WordPress/gutenberg/pull/49974
- https://github.com/WordPress/gutenberg/pull/50234

This PR adds the endpoint to Core and adds tests.

Trac ticket: https://core.trac.wordpress.org/ticket/58524

## Testing

To test the endpoint in the site editor, fire up this branch and head over to the site editor.

In your browser's console, switch to the network tab and run the following :

```js
await wp.apiFetch( { url: `/index.php?rest_route=%2Fwp%2Fv2%2Fglobal-styles%2F${ wp.data.select('core').__experimentalGetCurrentGlobalStylesId() }%2Frevisions&_locale=user` } );
```


`npm run test:php -- --filter WP_REST_Global_Styles_Revisions_Controller_Test`

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ramonjd committed:
Adds an endpoint that returns revisions for the global styles custom post.
Adds tests

ramonjd committed:
Rolled back unnecessary custom post changes

ramonjd committed:
Added rest endpoint to reset schema

ramonjd commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,311 @@
+		}
+
+		if ( rest_is_field_included( 'date', $fields ) ) {
+			$data['date'] = $item->post_date;
```
Just checking, should I also copy across the date formatting methods?

```php
		if ( in_array( 'date', $fields, true ) ) {
			$data['date'] = $this->prepare_date_response( $post->post_date_gmt, $post->post_date );
		}
```

See: https://github.com/WordPress/wordpress-develop/blob/82b6dddc066c5b5e3219eecc60c3051317de3d24/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L564-L564

ramonjd replied:
Assuming "yes" 😄

ramonjd commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,311 @@
+		}
+
+		if ( rest_is_field_included( 'date_gmt', $fields ) ) {
+			$data['date_gmt'] = $item->post_date_gmt;
```
Same as above

ramonjd commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,311 @@
+			$data['id'] = (int) $item->ID;
+		}
+
+		if ( rest_is_field_included( 'modified', $fields ) ) {
```
Here too

ramonjd commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,311 @@
+			$data['modified'] = $item->post_modified;
+		}
+
+		if ( rest_is_field_included( 'modified_gmt', $fields ) ) {
```
Also here

ramonjd committed:
generated fixtures via npm run test:php -- --filter WP_Test_REST_Schema_Initialization

ramonjd commented on tests/phpunit/tests/rest-api/rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,236 @@
+	 * @covers WP_REST_Global_Styles_Controller::register_routes
+	 */
+	public function test_register_routes() {
+		$routes = rest_get_server()->get_routes();
```
This is failing on CI (tests pass locally for me for some reason)

`TypeError: array_keys(): Argument #1 ($array) must be of type array, bool given in /var/www/tests/phpunit/includes/factory/class-wp-unittest-factory-for-thing.php:171
`

Any hints? Did I miss an import/registration somewhere 🤔

ramonjd commented on tests/phpunit/tests/rest-api/rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,236 @@
+		$routes = rest_get_server()->get_routes();
+		$this->assertArrayHasKey(
+			'/wp/v2/global-styles/(?P<parent>[\d]+)/revisions',
+			$routes,
```
This is failing on CI (tests pass locally for me for some reason)

`TypeError: array_keys(): Argument #1 ($array) must be of type array, bool given in /var/www/tests/phpunit/includes/factory/class-wp-unittest-factory-for-thing.php:171
`

Any hints? Did I miss an import/registration somewhere 🤔

ramonjd committed:
Adds prepare_date_response to parse and format the date in the response object

ramonjd committed:
tearing down after tests: deleting users

ramonjd committed:
post name should be different to the one created in global styles rest api test :)

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,334 @@
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param WP_REST_Request $request The request instance.
+	 *
+	 * @return WP_REST_Response|WP_Error
```
```suggestion
	 * @param WP_REST_Request $request The request instance.
	 * @return WP_REST_Response|WP_Error
```

mukeshpanchal27 reviewed (commented):
Thanks @ramonjd, Can you add ticket number in unit tests?

ramonjd committed:
Ticket no in tests
Spacing between comments

TimothyBJacobs commented on tests/phpunit/tests/rest-api/rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,252 @@
+	 * @doesNotPerformAssertions
+	 */
+	public function test_context_param() {
+		// Controller does not use get_context_param().
```
The controller does have `context` set, so this should get tests.

ramonjd replied:
Thanks for that.

I was looking at how other tests implement `test_context_param()` ([example](https://github.com/WordPress/wordpress-develop/blob/b0f085ece37288ff1e144474c29cc0184feeaf14/tests/phpunit/tests/rest-api/rest-revisions-controller.php#L109)), and they're testing the data in `$data['endpoints']` after an `OPTIONS` call. 

This controller doesn't have any other endpoint, e.g., `WP_REST_Server::CREATABLE` and so I'm not sure how to test it.  `$data['endpoints']` doesn't have any testable data. Any tips?

I'll add another `get_items` test to check that the correct fields are returned however.

ramonjd replied:
I've added a test `test_get_item_embed_context` to check the difference when sending context as an arg in the GET request

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+		$is_global_styles_user_theme_json = isset( $raw_config['isGlobalStylesUserThemeJSON'] ) && true === $raw_config['isGlobalStylesUserThemeJSON'];
+
+		if ( $is_global_styles_user_theme_json ) {
+			$user_theme_revisions = wp_get_post_revisions(
+				$parent->ID,
+				array(
+					'posts_per_page' => 100,
+				)
+			);
```
This endpoint should be paginated instead of hardcoding a limit of 100 and not supporting going any deeper.

spacedmonkey replied:
Massive +1

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+				}
+			}
+		}
+
```
If no revisions are found, should this not return a WP_Error with a 404 status.

ramonjd replied:
I need to refactor this method a bit to support pagination, and will model things on `WP_REST_Revisions_Controller->get_items`.

Looking at `WP_REST_Revisions_Controller->get_items`, it returns an empty array so I'm assuming we should follow that example rather than return an error for consistency.

ramonjd replied:
We might, however need a 404 if the `$is_global_styles_user_theme_json` test fails

spacedmonkey replied:
Seems the other get_items methods, do not return an error but an empty array for items.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+			return $parent;
+		}
+		$response                         = array();
+		$raw_config                       = json_decode( $parent->post_content, true );
```
`json_decode` can error or return null. There should be error handling here.

ramonjd replied:
~~I can switch to `wp_json_encode`, which handles errors~~

Ignore me. I'm tripping.

ramonjd replied:
Actually, sorry, I get what you mean. 😮‍💨 Will add some handling. Thank you!

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+	public function prepare_item_for_response( $post, $request ) {
+		$parent = $this->get_parent( $request['parent'] );
+		// Retrieves global styles config as JSON.
+		$raw_revision_config = json_decode( $post->post_content, true );
```
https://github.com/WordPress/wordpress-develop/pull/4606/files#r1238826104

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+		$parent = $this->get_parent( $request['parent'] );
+		// Retrieves global styles config as JSON.
+		$raw_revision_config = json_decode( $post->post_content, true );
+		$config              = ( new WP_Theme_JSON( $raw_revision_config, 'custom' ) )->get_raw_data();
```
This is only needed it `settings` and `styles` is returned. Only setup this up if these fields are requested, save some CPU cycles.

ramonjd replied:
Good call!

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -433,6 +434,16 @@ protected function prepare_links( $id ) {
+		if ( post_type_supports( $this->post_type, 'revisions' ) ) {
+			$revisions                = wp_get_latest_revision_id_and_total_count( $id );
+			$revisions_count          = ! is_wp_error( $revisions ) ? $revisions['count'] : 0;
+			$revisions_base           = sprintf( '/%s/%s/%d/revisions', $this->namespace, $this->rest_base, $id );
```
```suggestion
			$revisions_base           = sprintf( '/%s/%d/revisions', $base, $id );
```

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
```
```suggestion
```

`@return void` annotations aren't used in WP-Dev.

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+	}
+
+	/**
+	 * A direct copy of WP_REST_Revisions_Controller->prepare_date_response().
+	 * Checks the post_date_gmt or modified_gmt and prepare any post or
+	 * modified date for single post output.
```
```suggestion
	 * Checks the post_date_gmt or modified_gmt and prepare any post or
	 * modified date for single post output.
	 *
	 * Duplicates WP_REST_Revisions_Controller::prepare_date_response
```

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+		}
+
+		/*
+		 * The same check as WP_REST_Global_Styles_Controller->get_item_permissions_check.
```
```suggestion
		 * The same check as WP_REST_Global_Styles_Controller::get_item_permissions_check.
```

Nit. Sorry.

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,333 @@
+	}
+
+	/**
+	 * Get the parent post, if the ID is valid. Copied from WP_REST_Revisions_Controller.
```
```suggestion
	 * Get the parent post, if the ID is valid.
	 *
	 * Duplicate of WP_REST_Revisions_Controller::get_parent.
```

peterwilsoncc reviewed (commented):
A couple of nit picks, really.

ramonjd committed:
Adding feedback:
- Introducing pagination
- Updating tests

ramonjd committed:
Updated annotations that refer to duplicates

reformatting functions in test

ramonjd commented:
Thank @TimothyBJacobs @spacedmonkey and @peterwilsoncc for helping me make this patch better.

I think I've addressed all of the feedback, (except for the [context](https://github.com/WordPress/wordpress-develop/pull/4606#discussion_r1239245516) test) at least I think so. Please let me know where it can be improved.

ramonjd committed:
Returning error if there is one from prepare_item_for_response

ramonjd committed:
updated fixtures

ramonjd committed:
updated fixtures

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php:
```diff
@@ -254,7 +254,6 @@ public function get_items( $request ) {
 				'posts_per_page'   => -1,
 				'orderby'          => 'date ID',
 				'order'            => 'DESC',
-				'suppress_filters' => true,
```
Why is this being removed?

ramonjd replied:
oops, thanks for spotting! was testing something unrelated

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+				'post_parent'      => $parent->ID,
+				'post_type'        => 'revision',
+				'post_status'      => 'inherit',
+				'posts_per_page'   => -1,
```
-1 calls need be avoided at ALL COSTS. There is no case where this should be done in production. Please remove this line.

spacedmonkey replied:
Nevermind this was in the revisioins controller. Odd.

ramonjd replied:
Oh, could you explain why? 

Should it not also be here as well?

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L254

spacedmonkey replied:
It should not no. Unlimited queries can take database servers down if you are not careful.

ramonjd replied:
All good! Thanks for raising it anyway. I pinched it from the revisions controller.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$schema = array(
```
Why is the shape of this API response so different from global endpoint. Like title is missing.

ramonjd replied:
Global revisions is a bit of a special case, hence all the copying over work.  

It's one-off custom post created to only store user styles and can't be interacted with directly (only via the site editor UI).

The title cannot be edited and is irrelevant to the functionality.

ramonjd committed:
Reinstate suppress filter.
Was a mistake

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+	}
+
+	/**
+	 * Registers the controllers routes.
```
```suggestion
	 * Registers the controllers routes.
	 *
	 * @since 6.3.0
```

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+	 * @return array Collection parameters.
+	 */
+	public function get_collection_params() {
+		return array(
+			'context'  => array( 'default' => 'view' ),
+			'page'     => array(
+				'description'       => __( 'Current page of the collection.' ),
+				'type'              => 'integer',
+				'default'           => 1,
+				'sanitize_callback' => 'absint',
+				'validate_callback' => 'rest_validate_request_arg',
+				'minimum'           => 1,
+			),
+			'per_page' => array(
+				'description'       => __( 'Maximum number of items to be returned in result set.' ),
+				'type'              => 'integer',
+				'minimum'           => 1,
+				'maximum'           => 100,
+				'sanitize_callback' => 'absint',
+				'validate_callback' => 'rest_validate_request_arg',
+			),
+			'offset'   => array(
+				'description' => __( 'Offset the result set by a specific number of items.' ),
+				'type'        => 'integer',
+			),
+		);
```
```suggestion
		$collection_params = $this->get_collection_params();
		unset( $collection_params['search'] );
		
		return $collection_params;
```

This might be a little cleaner. And less translated strings.

ramonjd replied:
Looks good! I think I'll still need to add `offset` since it's not inheriting from `WP_REST_Revisions_Controller`

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+			$data['parent'] = (int) $parent->ID;
+		}
+
+		if ( rest_is_field_included( 'settings', $fields ) ) {
+			$data['settings'] = ! empty( $global_styles_config['settings'] ) ? $global_styles_config['settings'] : new stdClass();
+		}
+
+		if ( rest_is_field_included( 'styles', $fields ) ) {
+			$data['styles'] = ! empty( $global_styles_config['styles'] ) ? $global_styles_config['styles'] : new stdClass();
+		}
```
```suggestion
```

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,484 @@
+		}
+
+		if ( ! empty( $global_styles_config['styles'] ) || ! empty( $global_styles_config['settings'] ) ) {
+			$global_styles_config = ( new WP_Theme_JSON( $global_styles_config, 'custom' ) )->get_raw_data();
```
```suggestion
			$global_styles_config = ( new WP_Theme_JSON( $global_styles_config, 'custom' ) )->get_raw_data();
			if ( rest_is_field_included( 'settings', $fields ) ) {
			      $data['settings'] = ! empty( $global_styles_config['settings'] ) ? $global_styles_config['settings'] : new stdClass();
		        }

		      if ( rest_is_field_included( 'styles', $fields ) ) {
			     $data['styles'] = ! empty( $global_styles_config['styles'] ) ? $global_styles_config['styles'] : new stdClass();
		      }
```

ramonjd replied:
Nice, makes sense. Thanks!

spacedmonkey reviewed (changes requested):
A couple of nitpics. This is nearly there. Is there a committer / owner for this change?

ramonjd commented:
> A couple of nitpics. This is nearly there. Is there a committer / owner for this change?

I appreciate the continued feedback 🙇 

I'll get those changes in today and will liaise with @tellthemachines to get it committed

ramonjd committed:
Only run and set styles and/or settings through WP_Theme_JSON if one or both of them are set.

ramonjd committed:
forgot contect

ramonjd committed:
updated fixtures

ramonjd committed:
updated fixtures

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,470 @@
+	}
+
+	/**
+	 * Retrieves the query params for collections.
+	 * Inherits from WP_REST_Controller::get_collection_params(),
```
```suggestion
	 * Retrieves the query params for collections.
	 *
	 * Inherits from WP_REST_Controller::get_collection_params(),
```

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,470 @@
+	}
+
+	/**
+	 * Returns paginated revisions of the given global styles config custom post type.
+	 * The bulk of the body is taken from WP_REST_Revisions_Controller->get_items,
```
```suggestion
	 * Returns paginated revisions of the given global styles config custom post type.
	 *
	 * The bulk of the body is taken from WP_REST_Revisions_Controller->get_items,
```

peterwilsoncc commented on tests/phpunit/tests/rest-api/rest-global-styles-revisions-controller.php:
```diff
@@ -0,0 +1,805 @@
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/' . REST_TESTS_IMPOSSIBLY_HIGH_NUMBER . '/revisions' );
+		$response = rest_get_server()->dispatch( $request );
+		$this->assertErrorResponse( 'rest_post_invalid_parent', $response, 404 );
+	}
+	/**
```
```suggestion
	}

	/**
```

peterwilsoncc reviewed (commented):
A few nits but otherwise looks good.

ramonjd committed:
Update class-wp-rest-global-styles-revisions-controller.php

From feedback

ramonjd committed:
Update rest-global-styles-revisions-controller.php

Missing return

ramonjd committed:
linter!

spacedmonkey reviewed (approved):
This change looks good. The unit tests could do with `@covers`.

tellthemachines commented:
committed in [r56082](https://core.trac.wordpress.org/changeset/56082) / [c345180](https://github.com/WordPress/wordpress-develop/commit/c345180c29369eb11a53b9f55cbc9c19ae9e20f8).

ramonjd commented:
> The unit tests could do with @covers.

Thanks! PR for that here: https://github.com/WordPress/wordpress-develop/pull/4750
